binance-pipeline/lambdas/binance_historical_extraction.py

```diff
@@ -0,0 +1,249 @@
+        config['end_date'] = end_date
+
+    config['reverse'] = reverse
+
+    return config
+
+
+def create_payload(symbol: str, trade_id: int) -> dict:
+
+    return {"symbol": symbol, "limit": CHUNK_LIMIT, "fromId": trade_id}
+
+
+def send_sqs_message(queue_name: str, message: dict):
+    message = json.dumps(message)
+
+    # TO DO send the message to the SQS queue
+    #
+    #
+
+    return
+
+
+def handle_reverse_load(symbol: str, start_date: str = ''):
+    latest_trade_id = get_latest_trade_id(symbol)['id']
+
+    config = make_config(start_date, reverse=True)
+    orchestration_handler_generic(latest_trade_id, symbol, config)
+
+    return
+
+
+def handle_from_date(symbol: str, start_date: str, end_date: str = ''):
+    start_date = parse_date(start_date)
+
+    trade_id_finder = TradeIDFinder(symbol, start_date)
+    trade_id = trade_id_finder.get_trade_id_for_date()
+
+    config = make_config(start_date, end_date)
+    orchestration_handler_generic(trade_id, symbol, config)
+
+    return
+
+
+def handle_from_start(symbol, end_date: str = ''):
+    trade_id = 0 # assumed true
+
+    config = make_config(end_date)
+    orchestration_handler_generic(trade_id, symbol, config)
+
+    return
+
+
+def orchestration_handler_generic(trade_id, symbol, config):
+    reverse = config['reverse']
+    starting_ids = get_starting_ids(trade_id, RESERVED_CONCURRENCY, CHUNK_LIMIT, reverse)
+    for id_ in starting_ids:
+        payload = create_payload(symbol, id_)
+        config['payload'] = payload
+
+        message = {"config": config, "type": "api_call"}
+        send_sqs_message(SQS_QUEUE_NAME_EXTRACTION, message)
+
+
+def make_candidate_id(trade_id: int, delta: int, reverse: bool=False):
+    if reverse:
+        candidate_id = trade_id - delta
+    else:
+        candidate_id = trade_id + delta
+
+    return candidate_id
+
+
+def get_starting_ids(trade_id, concurrency, chunk_size, reverse=False):
+    starting_ids = []
+    for i in range(1, concurrency + 1):
+        delta = i*chunk_size
+        if (candidate_id := make_candidate_id(trade_id, delta, reverse)) >= 0:
+            starting_ids.append(candidate_id)
+
+    return starting_ids
+
+
+def handle_api_call(config: dict):
+    start = time.perf_counter()
+    payload = config['payload']
+
+    binance_query_api = BinanceQueryAPI()
+
+    status_code, res = binance_query_api.old_trade_lookup(
+                payload['symbol'], CHUNK_LIMIT, payload['fromId']
+            )
+
+    if status_code in [418, 429]:
+        handle_rate_limit(status_code, res, config)
+
+    # TO DO: store result to file
+    #
+    #
+
+    else:
+        delta = RESERVED_CONCURRENCY * CHUNK_LIMIT
+        next_trade_id = make_candidate_id(payload['fromId'], delta, config['reverse'])
+
+        if (len(res) == CHUNK_LIMIT) and (next_trade_id >= 0):
+            config['payload']['fromId'] = next_trade_id
+            message = {'config': config, 'type': 'api_call'}
+            end = time.perf_counter()
```

Comment:
Is there a context manager for `perf_counter`?

Reply:
one can be made quite simply, can we survive any more indentation? Perhaps a decorator on the method instead? If I add decorators to the methods I can actually sum them all when I reach the end of the main handler, as I do lose some small invocations by only targetting the longest running function. I am intending to underestimate the time by design, but I can afford to time the entire run time of the lambda and then deliberately skip the hot start time of the lambda env, and this would function in the same way as described in the README, which it doesn't strictly do currently